crates/accelerate/src/sparse_pauli_op.rs

```diff
@@ -257,10 +496,385 @@ fn decompose_dense_inner(
+            let mut values = Vec::<Complex64>::with_capacity(side * (num_ops + 1) / 2);
+            let mut indices = Vec::<$int_ty>::with_capacity(side * (num_ops + 1) / 2);
+            let mut indptr: Vec<$int_ty> = vec![0; side + 1];
+            let mut nnz = 0;
+            for i_row in 0..side {
+                order.sort_unstable_by(|&a, &b| {
+                    ((i_row as $uint_ty) ^ (paulis.x_like[a] as $uint_ty))
+                        .cmp(&((i_row as $uint_ty) ^ (paulis.x_like[b] as $uint_ty)))
+                });
+                let mut running = Complex64::new(0.0, 0.0);
+                let mut prev_index = i_row ^ (paulis.x_like[order[0]] as usize);
+                for (x_like, z_like, coeff) in order
+                    .iter()
+                    .map(|&i| (paulis.x_like[i], paulis.z_like[i], paulis.coeffs[i]))
+                {
+                    let coeff =
+                        if ((i_row as $uint_ty) & (z_like as $uint_ty)).count_ones() % 2 == 0 {
+                            coeff
+                        } else {
+                            -coeff
+                        };
+                    let index = i_row ^ (x_like as usize);
+                    if index == prev_index {
+                        running += coeff;
+                    } else {
+                        nnz += 1;
+                        values.push(running);
+                        indices.push(prev_index as $int_ty);
+                        running = coeff;
+                        prev_index = index;
+                    }
+                }
+                nnz += 1;
+                values.push(running);
+                indices.push(prev_index as $int_ty);
+                indptr[i_row + 1] = nnz;
+            }
+            (values, indices, indptr)
+        }
+
+        /// Build CSR data arrays for the matrix-compressed set of the Pauli operators, using a
+        /// parallel strategy.  This involves more data copying than the serial form, so there is a
+        /// nontrivial amount of parallel overhead.
+        fn $parallel_fn(paulis: &MatrixCompressedPaulis) -> CSRData<$int_ty> {
+            let side = 1 << paulis.num_qubits();
+            let num_ops = paulis.num_ops();
+            if num_ops == 0 {
+                return (vec![], vec![], vec![0; side + 1]);
+            }
+
+            let mut indptr = Vec::<$int_ty>::with_capacity(side + 1);
+            indptr.push(0);
+            // SAFETY: we allocate the space for the `indptr` array here, then each thread writes
+            // in the number of nonzero entries for each row it was responsible for.  We know ahead
+            // of time exactly how many entries we need (one per row, plus an explicit 0 to start).
+            // It's also important that `$int_ty` does not implement `Drop`, since otherwise it
+            // will be called on uninitialised memory (all primitive int types satisfy this).
```

Comment:
Do you have a way in mind to do this assertion?  The `macro_rules` is more like a template-match than a generic, so I'm not sure of an obvious way to put a "trait bound" on the input.

I guess I could possibly hack something together where we attempt to codegen something invalid if `$int_ty` or `$uint_ty` don't satisfy `PrimInt`, but I don't know if there's a better pattern.

Reply:
Not specifically I guess, I was thinking something like adding a dummy generic function that had the trait bound and passed `$int_ty` to it so we'd get a compile error. I'm not sure if there is a better pattern either TBH.

But, it's not important I was more just my idle musing. Especially since I don't think we'll ever make the macro use public so it's fine with just the comment.